hoomd/write/gsd.py

```diff
@@ -177,30 +177,34 @@ def _attach(self):
         super()._attach()
 
     @staticmethod
-    def write(state, filename, filter=All(), log=None):
+    def write(state, filename, filter=All(), mode='wb', log=None):
         """Write the given simulation state out to a GSD file.
 
         Args:
             state (State): Simulation state.
             filename (str): File name to write.
-            filter_ (``hoomd.ParticleFilter``): Select the particles to write.
-            log (``hoomd.logger.Logger``): A ``Logger`` object for GSD logging.
+            filter (`hoomd.ParticleFilter`): Select the particles to write.
+            mode (str): The file open mode. Defaults to `wb`.
+            log (`hoomd.logger.Logger`): Provide log quantities to write.
 
-        Note:
-            The file is always overwritten.
+        The valid file modes for `write` are ``'wb'`` and ``'xb'``.
```

Comment:
Why is append not allowed?

Reply:
This function is intended for use when writing single frames (i.e. a restart file at the end of a script). It would be inefficient and cause significant metadata I/O if users made repeated calls to `write` appending to the same file in place of using one writer operation that gets triggered (it keeps the file open).

Reply:
That's reasonable but not obvious to me. I am thinking about a use case that @ramanishsingh asked me about yesterday. He had a GSD file and needed to convert it to DCD for compatibility with some other tool. I wasn't aware of a good method to do that (garnett doesn't support writing DCD), so I helped him create a HOOMD v2 script that loaded the GSD system state frame-by-frame and wrote the DCD file that way. I had to do the same thing for a simulation a while ago where I needed the output of the HPMC SDF to compute a pressure, so I loaded the whole trajectory frame by frame and dumped the SDF. I could imagine wanting to do post-processing like that where HOOMD could write additional log quantities to the GSD file (adding a logger that wasn't originally present in the simulation run), because I do that explicitly for my current pathways analysis code, adding new metadata fields that cannot be computed on the fly (only after knowing the whole trajectory). I am not sure if the single-frame write would be appropriate for that, or if the standard writer approach allows for that.

Reply:
Conversion to a DCD file would be best achieved with a file format tool that can write DCD files. However, scanning through a GSD file's frames and recomputing HOOMD compute properties is something best handled by HOOMD itself. This came up in a recent mailing list discussion as well.

The single frame writer is not appropriate to write out multi-frame GSD files because of the overhead of opening and closing the file each time. Files like this should be written in append mode.

I propose a `GSDTrajectoryReplay` updater that reads frames from a GSD file in order and replaces the system state every time it is triggered. The canonical usage would run without an integrator and use this updater with a period of 1 so that each timestep corresponds to 1 frame in the GSD file. What do you think?

Reply:
The `GSDTrajectoryReplay` sounds like a great idea. I agree with all the comments about the other user needs that I mentioned. I felt there was something "related" about them which was kind of hard to put my finger on.

Reply:
I opened #823 to track this feature request.